sqlx-sqlite/src/index.rs

```diff
@@ -0,0 +1,599 @@
+            .await
+            .unwrap(); // TODO: handle error, possibly failing gracefully by scanning all files?
+
+        let mut file_scans: HashMap<String, (i64, ParquetAccessPlan)> = HashMap::new(); // file_name -> (file_size, row_groups)
+
+        for (file_name, file_size, file_row_group_counts, row_group_to_scan) in row_groups {
+            let (_, access_plan) = file_scans.entry(file_name).or_insert((
+                file_size,
+                ParquetAccessPlan::new_none(file_row_group_counts as usize),
+            ));
+            // Here we could do finer grained row-level filtering, but this example does not implement that
+            access_plan.set(row_group_to_scan as usize, RowGroupAccess::Scan)
+        }
+
+        Ok(file_scans
+            .into_iter()
+            .map(|(file_name, (file_size, access_plan))| {
+                (
+                    file_name,
+                    FileScanPlan {
+                        file_size: file_size as u64,
+                        access_plan,
+                    },
+                )
+            })
+            .collect())
+    }
+
+    /// Add a new file to the index
+    pub async fn add_file(&mut self, file: &Path) -> anyhow::Result<()> {
+        let file_name = file
+            .file_name()
+            .ok_or_else(|| internal_datafusion_err!("No filename"))?
+            .to_str()
+            .ok_or_else(|| internal_datafusion_err!("Invalid filename"))?;
+        let file_size = file.metadata()?.len();
+
+        let file = File::open(file).map_err(|e| {
+            DataFusionError::from(e).context(format!("Error opening file {file:?}"))
+        })?;
+
+        let reader = ParquetRecordBatchReaderBuilder::try_new(file)?;
+
+        // extract the parquet statistics from the file's footer
+        let metadata = reader.metadata();
+        let schema = reader.schema();
+        let parquet_schema = reader.parquet_schema();
+        let row_groups = metadata.row_groups();
+        let row_counts = StatisticsConverter::row_group_row_counts(row_groups.iter())?;
+        let mut row_group_statistics: Vec<_> = row_counts
+            .iter()
+            .enumerate()
+            .map(|(row_group, row_count)| {
+                RowGroupStatisticsInsert::new(row_group as i64, row_count.unwrap() as i64)
+            })
+            .collect();
+
+        for field in self.schema.fields() {
+            let column_name = field.name().clone();
+            let converter = StatisticsConverter::try_new(&column_name, schema, parquet_schema)?;
+            let min_values = converter.row_group_mins(row_groups.iter())?;
+            let max_values = converter.row_group_maxes(row_groups.iter())?;
+            let null_counts = converter.row_group_null_counts(row_groups.iter())?;
+            let null_counts = null_counts.as_primitive::<UInt64Type>();
+
+            for row_group in 0..metadata.num_row_groups() {
+                match field.data_type() {
```

Comment:
You could probably get some re-use here with two generic functions (one for `IntXType` and one for strings).

Reply:
Ok I nerd-sniped myself into trying this; https://github.com/davidhewitt/datafusion-async-parquet-index/commit/9fa9e07b1cbb87264377956adcac28d9587e77f6 - feel free to pull the commit if you want it :)

Reply:
I ended up going with e88641b which also means we do the downcasts once instead of for each value in the stats arrays. Let me know if that sounds good to you.

Reply:
Ah definitely better!